File: src/org/mentalsmash/whadl/model/UnitMember.java
package org.mentalsmash.whadl.model;

import org.mentalsmash.whadl.WhadlRuntimeException;
import org.mentalsmash.whadl.model.expressions.Expression;
import org.mentalsmash.whadl.model.expressions.LiteralExpression;
import org.mentalsmash.whadl.model.patterns.Pattern;

public class UnitMember extends BaseEntity {
	
	private Unit unit;
	
	private Pattern equipmentPattern = Pattern.DEFAULT_PATTERN;
	private Pattern superPattern = Pattern.DEFAULT_PATTERN;

	public UnitMember(String name) {
		super(name, new LiteralExpression(true));
	}
	
	public UnitMember(UnitMember mem) {
		super(mem.getName(),mem.getConditions());
		this.equipmentPattern = mem.equipmentPattern;
		this.superPattern = mem.superPattern;
		this.unit = mem.unit;
		this.setReference(mem.getReference());
	}

	public UnitMember(String name, Expression conditions) {
		super(name, conditions);
	}

	public Pattern getEquipmentPattern() {
		return equipmentPattern;
	}

	public void setEquipmentPattern(Pattern eq) {
		this.equipmentPattern = eq;
	}

	public static final String ATTRIBUTE_EQUIPMENT = "equipment";
	public static final String ATTRIBUTE_UNIT = "unit";

	
	@Override
	public Object getAttribute(String selector) {
		if (selector.equals(ATTRIBUTE_EQUIPMENT)) {
			return this.equipmentPattern;
		} if (selector.equals(ATTRIBUTE_UNIT)) {
			return this.equipmentPattern;
		} else {
			return super.getAttribute(selector);
		}
	}

	@Override
	public void setAttribute(String selector, Object value) {
		if (selector.equals(ATTRIBUTE_EQUIPMENT)) {
			throw new WhadlRuntimeException("Read-Only attribute: " + selector);
		} else {
			super.setAttribute(selector, value);
		}
	}

	public Pattern getSuperPattern() {
		return this.superPattern;
	}

	public void setSuperPattern(Pattern sup) {
		this.superPattern = sup;
	}

	public void setUnit(Unit unit) {
		this.unit = unit;
	}

	public Unit getUnit() {
		return unit;
	}

}

File: client.revised/js/mentalsmash/whadl/client/model/UnitMember.js
(function () {
	var model = WHADL.client.ns('model'),
		
	UnitMember = model.type('UnitMember', function (json) {
		var that = null,
		BaseEntity = model.require('BaseEntity', UnitMember),
		Pattern = model.require('Pattern');

		that = new BaseEntity(json.name, json);

		that.equipment = new Pattern(json.equipment);

		that.toString = function (indent, u) {
			var result = indent + "MEMBER: " + this.member.name + "\n";
			result += this.indent(indent, u) + "EQUIPMENT:\n";
			result += this.equipment.toString(this.indent(indent, u), u) + "\n";
			result += this.indent(indent, u) + "CONDITIONS: " +
						this.json.conditions + "\n";
			return result;
		};

		that.toHTML = function () {
			var result = '<div class="unit-member" id="' + this.name + '"><h5>' + this.name + "</h5>";

			result += '<div class="equipment">';
			result += '<h6>EQUIPMENT</h6>';
			result += this.equipment.toHTML();
			result += '</div>';

			if (this.json.conditions !== "true") {
				result += '<div class="conditions"><h6>CONDITIONS</h6><pre>';
				result += this.json.conditions;
				result += '</pre></div>';
			}

			result += '</div>';

			return result;
		};

		return that;
	});
}());